Platform/OpenNtfsDxe/Open.c

```diff
@@ -0,0 +1,378 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+extern INT64  mIndexCounter;
+
+EFI_STATUS
+EFIAPI
+FileOpen (
+  IN EFI_FILE_PROTOCOL   *This,
+  OUT EFI_FILE_PROTOCOL  **NewHandle,
+  IN CHAR16              *FileName,
+  IN UINT64              OpenMode,
+  IN UINT64              Attributes
+  )
+{
+  EFI_STATUS     Status;
+  EFI_NTFS_FILE  *File;
+  EFI_FS         *FileSystem;
+  EFI_NTFS_FILE  *NewFile;
+  CHAR16         Path[MAX_PATH];
+  CHAR16         CleanPath[MAX_PATH];
+  CHAR16         *DirName;
+  UINTN          Index;
+  UINTN          Length;
+  BOOLEAN        AbsolutePath;
+
+  ASSERT (This != NULL);
+  ASSERT (NewHandle != NULL);
+  ASSERT (FileName != NULL);
+
+  File         = (EFI_NTFS_FILE *)This;
+  FileSystem   = File->FileSystem;
+  AbsolutePath = (*FileName == L'\\');
+
+  ZeroMem (Path, sizeof (Path));
+
+  if (OpenMode != EFI_FILE_MODE_READ) {
+    DEBUG ((DEBUG_INFO, "NTFS: File '%s' can only be opened in read-only mode\n", FileName));
+    return EFI_WRITE_PROTECTED;
+  }
+
+  if ((*FileName == 0) || (CompareMem (FileName, L".", 2U) == 0)) {
```

Comment:
`*FileName == 0` is a bit strange; and better replace `2U` with `L_STR_LEN (L".")`

Reply:
Why 2U? This checks whether the name *starts* with “.”, not whether it *is* “.”. That’s why sizeof and the macros are useful :)